packages/react/src/file-upload/FileUpload.tsx

```diff
@@ -171,21 +194,79 @@ export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
 			invalid: invalid || !!errorSummary,
 		});
 
+		const valueChangeToken = JSON.stringify(
+			value.filter((file) =>
+				fileRejections.every(
+					(rejectedFile) => rejectedFile.file.path !== file.path
+				)
+			)
+		);
+
 		useEffect(() => {
-			setFileRejections(
-				dropzoneFileRejections.map(({ file, errors }) => ({
-					file,
-					errors: errors.map((error) => ({
-						code: error.code,
-						message: getFileRejectionErrorMessage(
-							error,
-							formattedMaxFileSize,
-							acceptedFilesSummary
-						),
-					})),
-				}))
-			);
-		}, [dropzoneFileRejections, formattedMaxFileSize, acceptedFilesSummary]);
+			const allAttemptedUploads = [
+				...value,
+				...fileRejections.map((rej) => rej.file),
+			];
+			const isOverMaxFilesLimit =
+				maxFiles && allAttemptedUploads.length > maxFiles;
+
+			if (isOverMaxFilesLimit) {
+				const tooManyFilesRejectionList: Array<RejectedFile> = value
+					.filter((_file, index) => index >= maxFiles)
```

Comment:
Rather than iterating through the entire list, could you just `slice` the array at the `maxFiles`?

```suggestion
					.slice(maxFiles)
```

Reply:
Good suggestion